File: src/main/scala/eightyDays/scala211/bank/Bank.scala
package eightyDays.scala211.bank

import eightyDays.scala211.bank.account.{Account, Amount}
import eightyDays.scala211.bank.partner.{Identification, Partner}

case class Bank(name: String, var partners: Map[Identification, Partner] = Map[Identification, Partner](), var accounts: Set[Account] = Set[Account]()) {
  def post(account: Account, value: Amount): (Account, Bank) = {
    // TODO fix with correct account method
    val updatedAccount = account.deposit(value)
    // TODO fix Try(updatedAccount)
    (updatedAccount.get, copy(accounts = accounts - account + updatedAccount.get))
  }

  def posts(posts: (Account, Amount)*) = posts.foldLeft((List[Account](), this)) { (accounts, post) =>
    val (updatedAccount, updatedBank) = accounts._2.post(post._1, post._2)
    (accounts._1 :+ updatedAccount, updatedBank)
  }

  def add(partnerId: Identification, accountFactory: (Partner => Account)): Account = {
    val newAccount = accountFactory(partners(partnerId))
    accounts = accounts + newAccount
    newAccount
  }

  def add(pPartner: Partner): (Identification, Bank) =
    partners
      .find(_._2 == pPartner)
      .map { entry => (entry._1, this) }
      .getOrElse {
        val result = Identification()
        (result, copy(partners = partners + (result -> pPartner)))
      }

  def find(predicate: Partner => Boolean): Option[Partner] = partners
    .find(p => predicate(p._2))
    .map(_._2)

  def filterPartners(predicate: Partner => Boolean): Set[Partner] =
    partners
      .filter(p => predicate(p._2))
      .values
      .toSet

  def filterAccounts(predicate: Account => Boolean): Set[Account] =
    accounts
      .filter(predicate)
}

File: steps/7/main/java/eightyDays/java8/Bank.java
package eightyDays.java8;

import eightyDays.java8.account.Account;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Collectors;

public class Bank {
    private final String name;
    private Map<Identification, Partner> partners = new HashMap<>();
    private Map<Identification, Account> accounts = new HashMap<>();

    public Bank(String pName) {
        name = pName;
    }

    public String getName() {
        return name;
    }

    public Optional<Partner> getPartner(Identification id) {
        return Optional.ofNullable(partners.get(id));
    }

    public Identification addPartner(Partner pPartner) {
        return partners
                .entrySet()
                .stream()
                .filter(entry -> entry.getValue().equals(pPartner))
                .findFirst()
                .map(Map.Entry::getKey)
                .orElseGet(() -> {
                    Identification newId = new Identification();
                    partners.put(newId, pPartner);
                    return newId;
                });
    }

    public Account openAccount(Identification pOwner, Function<Identification, Account> pFactoryMethod) {
        Account account = pFactoryMethod.apply(pOwner);
        accounts.put(account.getNumber(), account);
        return account;
    }

    protected Account post(Identification account, Function<Account, Account> action) {
        Account updatedAccount = action.apply(accounts.get(account));
        accounts.remove(account);
        accounts.put(account, updatedAccount);
        return updatedAccount;
    }

    public Account deposit(Identification pId, BigDecimal pValue) {
        return deposit(pId, pValue, LocalDateTime.now());
    }

    public Account deposit(Identification pId, BigDecimal pValue, LocalDateTime pValuta) {
        return post(pId, account -> account.deposit(pValue, pValuta));
    }

    public Account withdraw(Identification pId, BigDecimal pValue) {
        return withdraw(pId, pValue, LocalDateTime.now());
    }

    public Account withdraw(Identification pId, BigDecimal pValue, LocalDateTime pValuta) {
        return post(pId, account -> account.withdraw(pValue, pValuta));
    }

    public Optional<Account> getAccount(Identification pId) {
        return Optional.ofNullable(accounts.get(pId));
    }

    public BigDecimal getAssets(Identification pOwner) {
        return accounts.entrySet().stream()
                .filter(entry -> entry.getValue().getOwner().equals(pOwner))
                .map(entry -> entry.getValue().getBalance())
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    protected Set<Identification> getPartner(Predicate<Partner> predicate) {
        return partners.entrySet().stream()
                .filter(entry -> predicate.test(entry.getValue()))
                .map(Map.Entry::getKey)
                .collect(Collectors.toSet());
    }

    public Set<Identification> searchPartner(String pName) {
        return getPartner(partner -> partner.getName().equals(pName));
    }

    public Set<Identification> searchPerson(String pFirstname) {
        return getPartner(partner -> {
            if (partner instanceof Person) {
                return ((Person)partner).getFirstName().equals(pFirstname);
            } else {
                return false;
            }
        });
    }
}
